### DISCUSSION TITLE
Help/documentation about "cryptolib"

### DISCUSSION THREAD

**Discussion by nastiliano** [micropython/micropython#10537] [General]
Hello everyone,

I've been looking for documentation about "cryptolib" module and I can't find an explanation or examples about how to use the different AES modes and all about that.

Someone could you give me a link o some information (example or whatever) about an AES CTR string encryption/decryption for example?

Thank yo very much!!

**Comment by nastiliano** [comment]
> Hello everyone,
> 
> I've been looking for documentation about "cryptolib" module and I can't find an explanation or examples about how to use the different AES modes and all about that.
> 
> Someone could you give me a link o some information (example or whatever) about an AES CTR string encryption/decryption for example?
> 
> Thank yo very much!!

Hello everyone,

After hour and hour looking for some examples. I only found:

https://hwwong168.wordpress.com/2019/09/25/esp32-micropython-implementation-of-cryptographic/

import uos
from ucryptolib import aes
 
MODE_ECB = 1
MODE_CBC = 2
MODE_CTR = 6
BLOCK_SIZE = 16
key = b'I_am_32bytes=256bits_key_padding'

# ECB MODE
cipher = aes(key, MODE_ECB) 
plaintext = 'This is AES cryptographic'
print('Plain Text:', plaintext)
pad = BLOCK_SIZE - len(plaintext) % BLOCK_SIZE
plaintext = plaintext + " "*pad 
encrypted = cipher.encrypt(plaintext)
print('AES-ECB encrypted:', encrypted ) 
cipher = aes(key,1) # cipher has to renew for decrypt 
decrypted = cipher.decrypt(encrypted)
print('AES-ECB decrypted:', decrypted)

# CBC MODE
iv = uos.urandom(BLOCK_SIZE)
cipher = aes(key,MODE_CBC,iv) 
ct_bytes = iv + cipher.encrypt(plaintext)
print ('AES-CBC encrypted:', ct_bytes) 
iv = ct_bytes[:BLOCK_SIZE]
cipher = aes(key,MODE_CBC,iv)
decrypted = cipher.decrypt(ct_bytes)[BLOCK_SIZE:]
print('AES-CBC decrypted:', decrypted)

# CTR MODE
Could somebody put the CTR MODE example???  I tried it for hours and an error in  "syntax mode" is returned...

**Comment by jimmo** [comment]
The documentation is here: https://docs.micropython.org/en/latest/library/cryptolib.html

> Could somebody put the CTR MODE example??? I tried it for hours and an error in "syntax mode" is returned...

@nastiliano Can you post your example that gives the error?

**Reply by nastiliano** [reply]
SOLVED:

Simply CTR MODE is not supported yet on the ESP32.   Sorry....

Thank you everyone!!

PD: I'll implement the CBC MODE
